File: src/main/java/ru/atsutsiev/sd/refactoring/database/ProductsTable.java
package ru.atsutsiev.sd.refactoring.database;

import org.jetbrains.annotations.NotNull;
import ru.atsutsiev.sd.refactoring.database.product.Product;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.joining;

/**
 * @author atsutsiev
 */
public class ProductsTable {
    private static final String TABLE_NAME = "PRODUCT";
    private static final String AVAILABLE_QUERIES = "max, min, sum, count, all";
    private final String databaseURL;

    public ProductsTable(final String databaseURL) {
        this.databaseURL = databaseURL;
    }

    public String queryAsHTML(final String queryCommand) throws IllegalArgumentException, SQLException {
        switch (queryCommand) {
            case "all"  : return toHTML(all());
            case "sum"  : return toHTML("Summary price: ", sum());
            case "count": return toHTML("Number of products: ", count());
            case "max"  : {
                final String header = "Product with max price: ";
                return max().map(p -> toHTML(header, p)).orElseGet(() -> toHTML(header));
            }
            case "min"  : {
                final String header = "Product with min price: ";
                return min().map(p -> toHTML(header, p)).orElseGet(() -> toHTML(header));
            }
            default     : return "Products table: available queries: " + AVAILABLE_QUERIES;
        }
    }

    @SuppressWarnings("UnusedReturnValue")
    public int createTable() throws SQLException {
        try (final Connection c = DriverManager.getConnection("jdbc:sqlite:test.db")) {
            try (Statement stmt = c.createStatement()) {
                return stmt.executeUpdate(
                        "CREATE TABLE IF NOT EXISTS PRODUCT" +
                            "(ID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL," +
                            " NAME           TEXT    NOT NULL, " +
                            " PRICE          INT     NOT NULL)"
                );
            }
        }
    }

    @SuppressWarnings("UnusedReturnValue")
    public int insert(@NotNull final Product product) throws SQLException {
        return executeUpdate(
                "INSERT INTO " + TABLE_NAME + " " +
                "(NAME, PRICE) VALUES " +
                "(\"" + product.getName() + "\"," + product.getPrice() + ")"
        );
    }

    public Connection getConnection() throws SQLException { return DriverManager.getConnection(databaseURL); }

    /* Queries */

    public List<Product> all() throws SQLException {
        return parseQueryOutput(executeQuery("SELECT * FROM " + TABLE_NAME));
    }

    public Optional<Product> max() throws SQLException {
        return parseQueryOutput(executeQuery("SELECT * FROM " + TABLE_NAME + " ORDER BY PRICE DESC LIMIT 1"))
                .stream()
                .findFirst();
    }

    public Optional<Product> min() throws SQLException {
        return parseQueryOutput(executeQuery("SELECT * FROM " + TABLE_NAME + " ORDER BY PRICE LIMIT 1"))
                .stream()
                .findFirst();
    }

    public long sum() throws SQLException {
        return executeQuery("SELECT SUM(price) as sum FROM " + TABLE_NAME).getLong("sum");
    }

    public int count() throws SQLException {
        return executeQuery("SELECT COUNT(*) as cnt FROM " + TABLE_NAME).getInt("cnt");
    }

    /* Internals */

    private ResultSet executeQuery(final String sql) throws SQLException {
        try (final Statement stmt = getConnection().createStatement()) { return stmt.executeQuery(sql); }
    }

    private int executeUpdate(final String sql) throws SQLException {
        try (final Statement stmt = getConnection().createStatement()) { return stmt.executeUpdate(sql); }
    }

    private List<Product> parseQueryOutput(ResultSet rs) throws SQLException {
        final List<Product> result = new ArrayList<>();
        while (rs.next()) {
            result.add(new Product(rs.getString("name"), rs.getInt("price")));
        }
        return result;
    }

    private static String toHTML(final List<Product> products) {
        return "<html><body>\n" +
                products.stream().map(Product::toHTML).collect(joining("\n")) +
                "</body></html>\n";
    }

    private static String toHTML(final String header, @NotNull final Product product) {
        return "<html><body>\n" +
                "<h1>" + header + "</h1>\n" +
                product.toHTML() + "\n" +
                "</body></html>";
    }

    private static String toHTML(final String header, final Object info) {
        return String.format("<html><body>\n%s\n%s\n</body></html>", header, info);
    }

    private static String toHTML(final String header) {
        return "<html><body>\n" +
                "<h1>" + header + "</h1>\n" +
                "</body></html>";
    }
}

File: src/main/kotlin/ru/atsutsiev/sd/refactoring/database/ProductsTable.kt
package ru.atsutsiev.sd.refactoring.database

import ru.atsutsiev.sd.refactoring.database.product.Product
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*
import java.util.stream.Collectors.joining

/**
 * @author atsutsiev
 */
class ProductsTable(private val databaseURL: String) {
    @get:Throws(SQLException::class)
    val connection: Connection
        get() = DriverManager.getConnection(databaseURL)

    @Throws(IllegalArgumentException::class, SQLException::class)
    fun queryAsHTML(queryCommand: String?): String {
        return when (queryCommand) {
            "all"   -> toHTML(all())
            "sum"   -> toHTML("Summary price: ", sum())
            "count" -> toHTML("Number of products: ", count())
            "max" -> {
                val header = "Product with max price: "
                max().map { p: Product -> toHTML(header, p) }.orElseGet { toHTML(header) }
            }
            "min" -> {
                val header = "Product with min price: "
                min().map { p: Product -> toHTML(header, p) }.orElseGet { toHTML(header) }
            }
            else -> "Products table: available queries: $AVAILABLE_QUERIES"
        }
    }

    @Throws(SQLException::class)
    fun insert(product: Product): Int {
        return executeUpdate("INSERT INTO $TABLE_NAME (NAME, PRICE) VALUES (\"${product.name}\",${product.price})")
    }

    @Throws(SQLException::class)
    fun createTable(): Int {
        DriverManager.getConnection("jdbc:sqlite:test.db").use { c ->
            c.createStatement().use { s ->
                return s.executeUpdate(
                    "CREATE TABLE IF NOT EXISTS PRODUCT" +
                            "(ID INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL," +
                            " NAME           TEXT    NOT NULL, " +
                            " PRICE          INT     NOT NULL)"
                )
            }
        }
    }

    /* Queries */
    
    @Throws(SQLException::class)
    fun all(): List<Product> {
        return parseQueryOutput(executeQuery("SELECT * FROM $TABLE_NAME"))
    }

    @Throws(SQLException::class)
    fun max(): Optional<Product> {
        return parseQueryOutput(executeQuery("SELECT * FROM $TABLE_NAME ORDER BY PRICE DESC LIMIT 1"))
            .stream()
            .findFirst()
    }

    @Throws(SQLException::class)
    fun min(): Optional<Product> {
        return parseQueryOutput(executeQuery("SELECT * FROM $TABLE_NAME ORDER BY PRICE LIMIT 1"))
            .stream()
            .findFirst()
    }

    @Throws(SQLException::class)
    fun sum(): Long = executeQuery("SELECT SUM(price) as sum FROM $TABLE_NAME").getLong("sum")

    @Throws(SQLException::class)
    fun count(): Int = executeQuery("SELECT COUNT(*) as cnt FROM $TABLE_NAME").getInt("cnt")

    /* Internals */
    @Throws(SQLException::class)
    private fun executeQuery(sql: String): ResultSet = connection.createStatement().use { s -> return s.executeQuery(sql) }

    @Throws(SQLException::class)
    private fun executeUpdate(sql: String): Int = connection.createStatement().use { s -> return s.executeUpdate(sql) }

    @Throws(SQLException::class)
    private fun parseQueryOutput(rs: ResultSet): List<Product> {
        return ArrayList<Product>().apply {
            while (rs.next()) {
                add(Product(rs.getString("name"), rs.getInt("price").toLong()))
            }
        }
    }

    companion object {
        private const val TABLE_NAME = "PRODUCT"
        private const val AVAILABLE_QUERIES = "max, min, sum, count, all"

        private fun toHTML(header: String): String = "<html><body>\n<h1>$header</h1>\b</body></html>"

        private fun toHTML(header: String, product: Product): String =
            "<html><body>\n<h1>$header</h1>\n${product.toHTML()}\n</body></html>"

        private fun toHTML(header: String, info: Any): String = "<html><body>\n$header\n$info\n</body></html>"

        private fun toHTML(products: List<Product>): String {
            return """
                <html><body>
                ${products.stream().map(Product::toHTML).collect(joining("\n"))}
                </body></html>
                """.trimIndent()
        }
    }
}